### PR TITLE
cmake: LD: Specify the entry point in the linker scripts

### PR DESCRIPTION
**Pull request by SebastianBoe**
The entry point can and therefore should be set by linker
scripts. Whenever possible one should express things in the source
language, be it .c or .ld, and not in code generators or in the build
system.

This patch removes the flag -eCONFIG_KERNEL_ENTRY from the linker's
command line and replaces it with the linker script command

ENTRY(CONFIG_KERNEL_ENTRY)

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

### CODE REVIEW

**Review by andyross — APPROVED** [review]
No complaint about the patch itself, but note that this whole feature just isn't very useful.   The "entry point" is a feature of the ELF file format (well, and COFF and others too of course).  It doesn't have anything to do with actual system bootstrapping or OS startup, all of which is SoC-dependent.  Some of our platforms use the ELF entry point (e.g. qemu loads an ELF file directly, and the esp32 loader gets the start address out of the ELF file).  Some don't care.  Some need it to point to something specific.

I wouldn't cry if the whole thing was taken out of the global build structure and put into per-arch code.

**Comment by rgundi** [comment]
May i know why esp32 and intel_s1000 files have not been modified with this change? This is definitely applicable for intel_s1000 at least.

File: include/arch/posix/linker.ld [hunk 1]
```diff
@@ -21,6 +21,7 @@
 #include <linker/linker-defs.h>
 #include <linker/linker-tool.h>
 
+ENTRY(CONFIG_KERNEL_ENTRY) /* TODO: Necessary? */
```
[the remark this answers is not in the document]
Comment (SebastianBoe) [reply include/arch/posix/linker.ld]: Fixed.

**Comment by SebastianBoe** [comment]
> May i know why esp32 and intel_s1000 files have not been modified with this change?

This was an oversight, thank you for reporting.

I have fixed those two linker scripts and double-checked that all xtensa linker scripts have been ported.

**Comment by SebastianBoe** [comment]
@andyross : Thank you for the insight.

It is good to know that Zephyr does not use this in it's boot-sequence.

As you noted, it is still useful to consistently set it to something reasonable as various tooling believes that the "program" "starts" at the entry point, even though the precise meaning is blurred in embedded ...

**Review by rgundi — APPROVED** [review]
LGTM

**Review by kgugala — APPROVED** [review]

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**